### ISSUE TITLE
Peg-out request unlock script verification issue

### ISSUE THREAD

**Issue by MarvinJanssen** [Trust-Machines/core-eng#275]
From my understanding, the latest spec of the peg-out process is basically as follows:

1. An unlock script is generated that states "Here are 80 bytes but ignore them. Before height N, X can spend, and after height N, Y can spend." The 80 bytes contains a Stacks account identifier.
2. The unlock script is used to generate a P2TR address.
3. The user that wants to peg out funds the P2TR address with some Bitcoin, to be used by the protocol to pay for the Bitcoin transaction fee required to fulfil the peg-out request.
4. The transaction funding the P2TR address of (3) AND the unlock script of (1) are submitted to the protocol and stored in a Clarity smart contract.
5. The protocol participants observe the smart contract in (4) for new peg-out requests, and threshold-sign fulfilment transactions if the conditions are met.

The problem is in (4): the Clarity contracts gets a Bitcoin transaction sending some tokens to a P2TR address and the supposed script that will allow those tokens to be spent. **It will actually be quite difficult (or maybe even impossible) to check that the provided unlock script can actually spend the tokens**. If we cannot verify whether the unlock script is actually correct, it means that anyone could submit faulty or other random transactions as "peg-out requests" and mess with the protocol. At least for sBTC Mini, it means that the protocol can slowed down significantly, as it will only move to the next cycle after all peg-out requests of the cycle have been fulfilled. If the P2TR cannot be spent via the expected path, then it is left until it expires.

The witness program is a `OP_1` followed by a 32 byte tweaked pub key Q, which is calculated as such: `Q = P + hash(P||m)G`. We need the original pub key P as well as implement the BIP0341 script validation rules in Clarity, and we need uint256. ([Maybe we can use this?](https://github.com/KStasi/clarity-uint256-lib))

See also:
- https://en.bitcoin.it/wiki/BIP_0341#Script_validation_rules

**Comment by radicleart** [comment]
Does the problem in 4) depend on whether the tx fee payment is to a script hash or taproot pubkey directly ? I.e the difference between sending the stacks data in op_return vs sending it as part of the unlock script ?

**Comment by MarvinJanssen** [comment]
Two points:

1. For a P2WSH address, it is trivial to verify. The witness program is just a sha256 hash of the witness script.
2. `OP_RETURN` does not need to be verified in itself. If the transaction was mined then we only need to extract the data.

**Comment by jo-tm** [comment]
@MarvinJanssen @radicleart  can you please provide an update on this here? Is this still an issue?

**Comment by jo-tm** [comment]
@MarvinJanssen will check the Peg-Out examples with wrapping makes this easy. Still considered an issue.

**Comment by jo-tm** [comment]
Marvin: wrapping it is not gonna solve this issue. cc/ Jude

**Comment by jo-tm** [comment]
Steps according to Jude to handle this:

This section needs to include:

* It can verify that a threshold of stackers have voted to deny the peg-out request, and in doing so, burn the sBTC and trigger a BTC burn by the stackers (option 2)

* It can verify and tabulate votes from stackers that a given peg-out request is unsatisfiable, and maintain the lock on the requested sBTC until the start of the next reward cycle (option 3)

* It can determine that a peg-out request has not received the attention of at least 70% of the stackers by STX weight within a protocol-defined deadline (option 4)

**Comment by xoloki** [comment]
I really need to read the current mini and full sBTC docs.  But there's a couple of problems with what I'm seeing here.

In WSTS, there's no voting "no".  You either get a threshold to sign, or you don't.  We could consider bad signatures as voting "no" if we want to go that route.  

Secondly, why are we requiring people to pay btc fees to fund btc fees?  We're essentially requiring people to pay twice for a peg out.  And if we actually can't verify that the peg out represents a valid STX spend, that's really bad.

It seems to me that it makes much more sense for peg out to be a STX tx, sending sBTC to a STX peg out address, with a memo field saying which BTC address to send the peg out.  This way there's no question it's a valid sBTC spend.  Then the BTC tx fees come from the BTC being pegged out.

Does this make sense?

**Comment by jo-tm** [comment]
@xoloki @igor is going to ping you on the context, he says that to refuse you can sign a message refusing. thanks

**Comment by MarvinJanssen** [comment]
@friedger here is the issue tracking taproot script verification. Can you share your thoughts?

**Comment by jo-tm** [comment]
@MarvinJanssen: We need to do more work on this, still an open question.
@friedger pls add any thoughts on this issue.

**Comment by jo-tm** [comment]
@friedger : will give an opinion on this Risk issue.

**Comment by friedger** [comment]
> as it will only move to the next cycle after all peg-out requests of the cycle have been fulfilled

Jude suggested that the specification is changed so that some peg-out requests can be invalidated for the current cycle.
Would we implement all of the options 2 - 4 as well?

**Comment by netrome** [comment]
I'm late to the party, but if this is still an issue I wonder why the peg out processor even have to consider commit transactions. Because if I understand the problem here, it's that it's difficult to verify that a provided script can spend the tokens. However, why do we even need to check that? Wouldn't it be better if the contract only read the reveal (or acknowledge as it's called in the old Master sBTC doc) transaction. Afaik that's how peg-ins/deposits work.

I would therefore imagine something like this:
1. User commits to their withdrawal tx.
2. Signers acknowledge and sign the withdrawal reveal tx.
3. User posts this tx to the peg-out processor, burning (or locking) their sBTC.
4. Signers fulfill the withdrawal. 

Would there be any problems with this flow?

**Comment by MarvinJanssen** [comment]
Sure that would work, but the design doc requires that the sBTC protocol becomes aware of all pending redeem/withdrawal requests. There is no way for the protocol to know if signers are eschewing their duty otherwise. (And they cannot be penalised.)

**Comment by netrome** [comment]
> Sure that would work, but the design doc requires that the sBTC protocol becomes aware of all pending redeem/withdrawal requests. There is no way for the protocol to know if signers are eschewing their duty otherwise. (And they cannot be penalised.)

How is that handled for deposits? Or do we suffer from the same potential spam problem there?

**Comment by MarvinJanssen** [comment]
We do not track deposits in the same way. People can fund P2TR escrow addresses and when and if those get picked up by the signers, those transactions are used as proof to mint sBTC. This difference exist because the protocol does not need to be aware of pending deposit requests.

**Comment by netrome** [comment]
But the signers could still censor deposits by not revealing pending deposits. In which way is that not eschewing their duty in the same way as not revealing pending withdrawals?

**Comment by MarvinJanssen** [comment]
In that case the end-user's capital is not at risk. So people take their Bitcoin back and they likely won't ever try to use sBTC again.

**Comment by FriendsFerdinand** [comment]
I think it's likely that's we're going to be initiating peg-outs from the Stacks chain using sBTC.

Though, the option to force a peg-out by initiating a Bitcoin transaction was mentioned. I think it's fair to assume that in that case, the script would have to be revealed before we can force signers to accept the peg-out request.

**Comment by netrome** [comment]
Yeah, given the points raised in https://github.com/stacks-network/sbtc-docs/issues/49 we'll likely keep aiming at having the primary sBTC interface being with Bitcoin operations for 1.0. To keep the scope smaller, I don't see us supporting Stacks-initiated sBTC transactions for 1.0 but it could be a nice feature for 1.1.
